packages/lesswrong/components/users/DialogueMatchingPage.tsx

```diff
@@ -353,41 +630,247 @@ const UserTopTags = ({ classes, targetUserId }: { classes: ClassesType, targetUs
+  const [topicPreferences, setTopicPreferences] = useState<ExtendedDialogueMatchPreferenceTopic[]>(Object.values(mergedTopicDict))
+
+  useEffect(() => setTopicPreferences(topicPreferences => {
+    const existingTopicDict = Object.fromEntries(topicPreferences.map(topic => [topic.text, topic]))
+    const newRecommendedTopicDict = Object.fromEntries(topicRecommendations?.map(comment => [comment.contents?.plaintextMainText ?? '', {
+      text: comment.contents?.plaintextMainText ?? '',
+      preference: 'No' as const,
+      commentSourceId: comment._id
+    }]) ?? [])
+    const mergedTopicDict = mergeWith(existingTopicDict, newRecommendedTopicDict, (existingTopic, newRecommendedTopic) => ({...newRecommendedTopic, ...existingTopic}))
+    return Object.values(mergedTopicDict)
+  }), [topicRecommendations])
+
+  useEffect
+
+  const [recommendedTopics, userSuggestedTopics]  = partition(topicPreferences, topic => topic.matchedPersonPreference !== "Yes")
+  if (called && !loadingCreatedMatchPreference && !(newMatchPreference as any)?.createDialogueMatchPreference?.data?.generatedDialogueId) {
+    return (
+      <LWDialog open onClose={onClose}>
+          <DialogTitle>
+            <h2>You submitted, nice job.</h2>
+            <p>This info will be sent to your match partner.</p> 
+            <p>Once they fill out the form, you'll get to see each other's answers and chat about whether a dialogue makes sense.</p>
+          </DialogTitle>
+          <div style={{textAlign: "center"}}>
+            <img style={{maxHeight: "50px"}} src="https://res.cloudinary.com/lesswrong-2-0/image/upload/v1497915096/favicon_lncumn.ico"></img>
+          </div>
+          <DialogActions>
+            <Button onClick={onClose} color="default">
+              Close
+            </Button>
+          </DialogActions>
+      </LWDialog>
+  )}
+
+  return (
+    <LWDialog open onClose={onClose}>
+      <div className={classes.dialogBox}>
+        <DialogTitle className={classes.dialogueTitle}>Alright, you matched with {targetUserDisplayName}!</DialogTitle>
+        <DialogContent >
+          {userSuggestedTopics.length > 0 && <>
+            <div>Here are some topics {targetUserDisplayName} was interested in:</div>
+            <div className={classes.dialogueTopicList}>
+              {userSuggestedTopics.map((topic) => <div className={classes.dialogueTopicRow} key={topic.text}>
+                <Checkbox 
+                    className={classes.dialogueTopicRowTopicCheckbox}
+                    color={topic.matchedPersonPreference === "Yes" ? "primary" : "default"}
+                    checked={topic.preference === "Yes"}
+                    // Set the preference of the topic with the matching text to the new preference
+                    onChange={event => {
+                      const topicToUpdateIx = topicPreferences.findIndex(({text}) => text === topic.text)
```

Comment:
@b0b3rt I refactored this code away from the version using a `map` with a complicated inner function over all the topics. I think this resolves your previous comment in spirit